### DISCUSSION TITLE
provide a function (with Option-Enum-Result) to convert a n int to enum

### DISCUSSION THREAD

**Discussion by WoodyAtHome** [vlang/v#18601] [Ideas]
I think int to enum conversion is still one of the weakest point in V. Converting Enums to Int  Conversion happens very often and it should be always checked if the int has a valid enum value. And if not, the case has to be handled. V provides at this point the excellent option results.

So I think  V urgently needs a function to convert Ints to Enums in this style:
```
enum MyEnum {
  a = 1
  b = 5
  c = 7
}

fn main() {
  en1 := convert_enum(MyEnum, 1) or {panic('1 is not v valid MyEnum Value')} // prints no Output here
  dump(en1) // prints MyEnum.a
  en2 := convert_enum(MyEnum, 2) or {panic('2 is not v valid MyEnum Value')} // prints '2 is not v valid MyEnum Value'
}
```

**Comment by SurmanPP** [comment]
That's much better than enforcing unsafe imo. As just throwing unsafe at things does not improve safety at all. Instead possible errors should be handled.

**Comment by medvednikov** [comment]
I like the idea.

@spytheman what do you think?

**Comment by Delta456** [comment]
It seems fine with me too. The syntax can be `Enum(int) or {}`, just like we have for array and map indexes.

**Comment by JalonSolov** [comment]
`Enum(int)` would be a cast, not a function.

**Comment by spytheman** [comment]
If we had static methods, we could automatically generate:
`fn MyEnum.from_int(input int) ?MyEnum` 
and 
`fn MyEnum.from_string(input string) ?MyEnum`

**Comment by lancethepants** [comment]
```
enum Color {
	red = 0
	orange
	yellow
	green
	blue
	purple
}

color = Color(5)
```
Yes, would also like to see int -> enum casting without unsafe{} block.

**Comment by JalonSolov** [comment]
int -> enum will _always_ be unsafe, as you can't guarantee an enum with that int value exists.

**Comment by SurmanPP** [comment]
If you have an or block you can handle exceptions and don't need to make it unsafe. Or am I wrong?

**Comment by spytheman** [comment]
```v
fn enumerator<T>(x int) !T {
    z := unsafe { T(x) }
    if z.str() == 'unknown' {
        return error('invalid enum value x: $x')
    }
    return z
}

enum MyEnum1 { a = 0 b c }
enum MyEnum2 { a = 10 b c }
fn main() {
    dump( enumerator<MyEnum1>(1)! )
    dump( enumerator<MyEnum2>(1)! )
}
```
this works, but has the disadvantage that `if z.str() == 'unknown' {` is slow, and it will break for enums with a symbol value `unknown`
